compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -286,6 +288,86 @@ pub enum Visibility<Id = LocalDefId> {
     Restricted(Id),
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Copy, Hash, Encodable, Decodable, HashStable)]
+pub enum Restriction<Id = DefId> {
+    /// The restriction does not affect the item.
+    Unrestricted,
+    /// The restriction only applies outside of this path.
+    Restricted(Id, Span),
+}
+
+impl<Id: Into<DefId> + Copy> Restriction<Id> {
+    /// Returns `true` if this restriction applies in the given module. This
+    /// means the behavior is _not_ allowed.
+    pub fn is_restricted_in(self, module: DefId, tcx: TyCtxt<'_>) -> bool {
+        let restricted_to = match self {
+            Restriction::Unrestricted => return false,
+            Restriction::Restricted(module, _) => module,
+        };
+
+        !tcx.is_descendant_of(module, restricted_to.into())
+    }
+
+    /// Obtain the [`Span`] of the restriction. Panics if the restriction is unrestricted.
+    pub fn expect_span(&self) -> Span {
+        match self {
+            Restriction::Unrestricted => bug!("called `expect_span` on an unrestricted item"),
+            Restriction::Restricted(_, span) => *span,
+        }
+    }
+
+    pub fn expect_restriction_path(
+        &self,
+        tcx: TyCtxt<'_>,
+        krate: rustc_span::def_id::CrateNum,
+    ) -> String {
+        let Restriction::Restricted(def_id, _) = self else {
+            bug!("called `expect_restricted_path` on an unrestricted item")
+        };
+
+        let def_id: DefId = (*def_id).into();
+
+        if krate == def_id.krate {
+            tcx.def_path_str(def_id)
+        } else {
+            tcx.crate_name(def_id.krate).to_ident_string()
+        }
+    }
+
+    /// Obtain the stricter of the two restrictions. If the two restrictions are the same, returns
+    /// `left`. Panics if the restrictions do not reference the same crate.
+    pub fn stricter_of(left: Self, right: Self, tcx: TyCtxt<'_>) -> Self {
+        match (left, right) {
+            (Restriction::Unrestricted, Restriction::Unrestricted) => Restriction::Unrestricted,
+            (Restriction::Unrestricted, Restriction::Restricted(..)) => right,
+            (Restriction::Restricted(..), Restriction::Unrestricted) => left,
+            (Restriction::Restricted(left_did, _), Restriction::Restricted(right_did, _)) => {
+                let left_did = left_did.into();
+                let right_did = right_did.into();
+
+                if left_did.krate != right_did.krate {
+                    bug!("stricter_of: left and right restriction do not reference the same crate");
+                }
+
+                if tcx.is_descendant_of(left_did, right_did) { left } else { right }
```

Comment:
isn't this weird in case `left`/`right` are siblings?

Reply:
I'm not sure what you mean. If you mean they share the same parent module, the crate is still the same.

Reply:
I'm not certain what you mean. Given that restrictions are limited to the current module or ancestor modules, how could left and right be siblings? Assuming that by siblings you mean two modules with the same parent.